File: backend/app/use_cases/video/youtube.py
from __future__ import annotations

from urllib.parse import parse_qs, urlparse

from app.use_cases.video.exceptions import InvalidYoutubeUrl


_VALID_HOSTS = {
    "youtube.com",
    "www.youtube.com",
    "m.youtube.com",
    "youtu.be",
    "www.youtu.be",
}


def extract_youtube_video_id(url: str) -> str:
    parsed = urlparse(url.strip())
    if parsed.scheme not in {"http", "https"} or parsed.netloc.lower() not in _VALID_HOSTS:
        raise InvalidYoutubeUrl("Invalid YouTube URL.")

    host = parsed.netloc.lower()
    if host.endswith("youtu.be"):
        candidate = parsed.path.strip("/").split("/")[0]
    elif parsed.path == "/watch":
        candidate = parse_qs(parsed.query).get("v", [""])[0]
    elif parsed.path.startswith("/embed/") or parsed.path.startswith("/shorts/"):
        candidate = parsed.path.strip("/").split("/")[1]
    else:
        candidate = ""

    if len(candidate) != 11 or not candidate.replace("-", "").replace("_", "").isalnum():
        raise InvalidYoutubeUrl("Invalid YouTube URL.")
    return candidate


def build_youtube_embed_url(video_id: str) -> str:
    return f"https://www.youtube.com/embed/{video_id}"

File: backend-hono/src/lib/youtube.ts
/**
 * YouTube URL 検証・動画 ID 抽出（use_cases/video/youtube.py extract_youtube_video_id と一致）。
 */
const VALID_HOSTS = new Set([
  "youtube.com",
  "www.youtube.com",
  "m.youtube.com",
  "youtu.be",
  "www.youtu.be",
]);

export const INVALID_YOUTUBE_URL_MESSAGE = "Invalid YouTube URL.";

/**
 * URL から 11 桁の動画 ID を抽出。不正なら null（呼び出し側で "Invalid YouTube URL."）。
 * Python urlparse + parse_qs のロジックを踏襲。
 */
export function extractYoutubeVideoId(url: string): string | null {
  let parsed: URL;
  try {
    parsed = new URL(url.trim());
  } catch {
    return null;
  }

  const scheme = parsed.protocol.replace(/:$/, "");
  const netloc = parsed.host.toLowerCase(); // host は hostname[:port]（Python netloc 相当）
  if ((scheme !== "http" && scheme !== "https") || !VALID_HOSTS.has(netloc)) {
    return null;
  }

  const path = parsed.pathname;
  let candidate = "";
  if (netloc.endsWith("youtu.be")) {
    candidate = path.replace(/^\/+|\/+$/g, "").split("/")[0] ?? "";
  } else if (path === "/watch") {
    candidate = parsed.searchParams.get("v") ?? "";
  } else if (path.startsWith("/embed/") || path.startsWith("/shorts/")) {
    candidate = path.replace(/^\/+|\/+$/g, "").split("/")[1] ?? "";
  }

  const stripped = candidate.replace(/[-_]/g, "");
  if (candidate.length !== 11 || stripped.length === 0 || !/^[a-zA-Z0-9]+$/.test(stripped)) {
    return null;
  }
  return candidate;
}
